README.txt
= rrs_fetcher

* http://gitorious.org/rr2fetcher

== DESCRIPTION:

RapidShare downloader

== FEATURES/PROBLEMS:

* unknown

== SYNOPSIS:

  FIX (code sample of usage)

== REQUIREMENTS:

* hpricot

== INSTALL:

$ sudo gem install hpricot --no-ri --no-rdoc
$ git clone git://gitorious.org/~krawek/rr2fetcher/krawek.git rrsfetcher
$ cd rrsfetcher
$ rake gem
$ sudo gem install --no-ri --no-rdoc pkg/*.gem
$ rrsfetcher links_or_file-with-list-of-urls


== LICENSE:

(The MIT License)

Copyright (c) 2009 FIX

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 0.0.1 2009-06-08

* 1 major enhancement:
  * Initial release

== 0.2 2009-07-07
* new release


test.rb
require 'rrsfetcher'
require 'benchmark'

# ff = page.search("#ff")
# action = ff.attr("action")
# name = "dl.start"
# value = "free"

File.open("fixtures/page2.html", "r") do |file|
  bnk = Benchmark.realtime do
    RRSFetcher.parse_second_page(file.read)
  end
  p bnk
end

lib/rr2fetcher.rb
class RR2Fetcher
  VERSION = "0.2"

  attr_accessor :download_list
  attr_accessor :app

  def initialize(options)
    @download_list = []
    @app = options[:backend]
    @options = options
  end

  def add_download(*links)
    links.each do |link|
      if File.exist?(link)
        File.open(link) do |f|
          @download_list += f.readlines
        end
      else
        @download_list << link
      end
    end
  end

  def start_dowload
    while !@download_list.empty?
      url = @download_list.pop.strip
      next if url.empty? || downloaded?(url)
      download url
    end
  end

  def download(to_download)
    parser = RSSParser.new(to_download)

    wait_until_ready(parser)
    agent = "Mozilla Firefox 1.5"

    cmd = "wget -U '#{agent}' '#{parser.download_link}' --post-data='mirror=on&x=67&y=50'"
    file = parser.filename # FIXME: check if download_link is valid

    case @app
      when :curl
        cmd = "curl -A '#{agent}' --url '#{parser.download_link}' --data 'mirror=on&x=67&y=50' > #{file}"
      when :kget4
        url = %@'#{parser.download_link}?mirror=on&x=67&y=50'@
        cmd = "qdbus org.kde.kget /kget/MainWindow_1 org.kde.kget.addTransfer '#{parser.download_link}' ./#{file} true"
    end

    system(cmd)
  end

  private
  def downloaded?(url)
    filename = url.to_s.split("/").last

    if File.exist?(filename)
      $stderr.puts "Already downloaded: #{filename}"
      return true
    end
    false
  end

  def wait_until_ready(parser)
    while parser.metadata.include?(:minutes)
      show_delay(parser.metadata[:minutes] * 60)
      parser.parse_next_page
    end

    show_delay(parser.metadata[:seconds])
  end

  def show_delay(seconds)
    seconds += 5

    pbar = ProgressBar.new("Waiting", seconds)
    seconds.times do |i|
      pbar.inc
      sleep(1)
    end
    pbar.finish
  end
end


lib/rrs_parser.rb
class RSSParser
  attr_reader :page, :metadata, :url, :download_link, :filename, :form_action
  def initialize(url)
    @page = Hpricot(Net::HTTP.get(URI.parse(url)))
    @url = url
    @download_link = ""
    @filename = ""

    find_form_action
  end

  def parse_next_page
    @metadata = {}

    resp = Net::HTTP.post_form(URI.parse(@form_action), {"dl.start"=>"free"})
    body = resp.body

    body.each_line do |line|
      if line =~ /.*name="dlf"\saction="(\S*)"/
        $stderr.puts "-> Downloading #{$1}..."
        @download_link = $1
        @filename = $1.to_s.split("/").last
      end

      if line =~ /var\s*c\s*=\s*(\d+)/
        @metadata[:seconds] = $1.to_i
      elsif line =~ /try\s*again\s*in\s*about\s*(\d+)\s*minutes/
        @metadata[:minutes] = $1.to_i
      end
    end

    @metadata
  end

  protected
  def find_form_action
    ff = @page.search("#ff")
    unless ff
      $stderr.puts "there was an error"
      return false
    end

    @form_action = ff.attr("action") # FIXME: Check the error/exception
    parse_next_page
    true
  end
end


lib/rrs_fetcher.rb
class RRSFetcher
  VERSION = "0.1"

  attr_accessor :download_list
  attr_accessor :app

  def initialize(options)
    @download_list = []
    @app = options[:backend]
    @options = options
  end

  def add_download(*links)
    links.each do |link|
      if File.exist?(link)
        File.open(link) do |f|
          @download_list += f.readlines
        end
      else
        @download_list << link
      end
    end
  end

  def start_dowload
    while !@download_list.empty?
      url = @download_list.pop.strip
      next if url.empty? || downloaded?(url)
      download url
    end
  end

  def download(to_download)
    parser = RSSParser.new(to_download)

    wait_until_ready(parser)
    agent = "Mozilla Firefox 1.5"

    cmd = "wget -U '#{agent}' '#{parser.download_link}' --post-data='mirror=on&x=67&y=50'"
    file = parser.filename # FIXME: check if download_link is valid

    case @app
      when :curl
        cmd = "curl -A '#{agent}' --url '#{parser.download_link}' --data 'mirror=on&x=67&y=50' > #{file}"
      when :kget4
        url = %@'#{parser.download_link}?mirror=on&x=67&y=50'@
        cmd = "qdbus org.kde.kget /kget/MainWindow_1 org.kde.kget.addTransfer '#{parser.download_link}' ./#{file} true"
    end

    system(cmd)
  end

  private
  def downloaded?(url)
    filename = url.to_s.split("/").last

    if File.exist?(filename)
      $stderr.puts "Already downloaded: #{filename}"
      return true
    end
    false
  end

  def wait_until_ready(parser)
    while parser.metadata.include?(:minutes)
      show_delay(parser.metadata[:minutes] * 60)
      parser.parse_next_page
    end

    show_delay(parser.metadata[:seconds])
  end

  def show_delay(seconds)
    seconds += 5

    pbar = ProgressBar.new("Waiting", seconds)
    seconds.times do |i|
      pbar.inc
      sleep(1)
    end
    pbar.finish
  end
end


rrsfetcher.rb
#!/usr/bin/env ruby 
require 'rubygems'
require 'hpricot'
require 'net/http'
require 'uri'
# require 'open3'


class RRSFetcher
  attr_accessor :download_list
  attr_accessor :app
  def initialize
    @download_list = []
    @app = :wget
  end

  dee self.get_page(url)
    Hpricot(Net::HTTP.get(URI.parse(url)))
  end

  def self.parse_second_page(body)
    rest = {}
    body.each_line do |line|
      if line =~ /.*name="dlf"\saction="(\S*)"/
        rest[:dlf] = $1
      end

      if line =~ /var\s*c\s*=\s*(\d+)/
        rest[:seconds] = $1.to_i
      elsif line =~ /try\s*again\s*in\s*about\s*(\d+)\s*minutes/
        rest[:minutes] = $1.to_i
      end
    end
    rest
  end

  def self.link_propieties(link)
    page = RRSFetcher.get_page(link)
    ff = page.search("#ff")
    unless ff
      #DEBUG
      File.open("error_page.html", "wb") do |file|
        file.write(page)
      end
    else
      action = ff.attr("action")
      name = "dl.start"
      value = "free"
      resp = Net::HTTP.post_form(URI.parse(action), {name=>value})
      result = RRSFetcher.parse_second_page(resp.body)
      result[:action] = action
      result
    end
  end

  def add_download( link)
    p "adding #{link} to download queue"
    @download_list << link
  end

  def start_dowload
    def show_delay(seconds)
      seconds.times do |i|
        print "#{i} ";
        sleep(1)
      end
    end
    while(!download_list.empty?)
      to_download = @download_list.pop
      propieties = RRSFetcher.link_propieties(to_download)
      while(propieties.include?(:minutes))
        p "minutes #{result[:minutes]}"
        show_delay(result[:minutes] * 60)
        propieties = Net::HTTP.post_form(URI.parse(action), {name=>value})
        RRSFetcher.parse_second_page(resp)
      end
      show_delay(propieties[:seconds])
      p "go!!"
      cmd = "wget #{propieties[:dlf]} --post-data='mirror=on&x=67&y=50'"
      file = propieties[:dlf].split("/").last
      case @app
      	when :curl
	  cmd = "curl --url #{propieties[:dlf]} --data 'mirror=on&x=67&y=50' > #{file}"
	when :wget
          cmd = "wget #{propieties[:dlf]} --post-data='mirror=on&x=67&y=50'"
	when :kget4
	  url = %@#{propieties[:dlf]}?mirror=on&x=67&y=50@
	  cmd = "qdbus org.kde.kget /kget/MainWindow_1 org.kde.kget.addTransfer '#{url}' ./#{file} true"
      end

      system(cmd)

#       process = Open3.popen3(cmd)
    end
  end
end

if __FILE__ == $0
  downloader = RRSFetcher.new
  downloader.app = :curl
  if ARGV[0]
    downloader.add_download(ARGV[0])
    thread = Thread.new do
      downloader.start_dowload
    end
    thread.join
  end
end
